libfreerdp/core/gateway/tsg.c

```diff
@@ -1920,7 +1920,7 @@ static int transport_bio_tsg_gets(BIO* bio, char* str, int size)
 
 static long transport_bio_tsg_ctrl(BIO* bio, int cmd, long arg1, void* arg2)
```

Comment:
It makes sense to implement BIO_C_GET_FD here as well.

Reply:
I'd expect it to be called when RDG RPC transport is used (/gt:rpc). That single BIO_get_fd call (from tls.c) should end up here in such case.

Reply:
@p-pautov I'd expect the same. But please test it! I did, and I asked to test for a reason... (e.g. things might work differently on other OSs or clients.)

My observations on this:
- Returning -1 would make the test fail anyway, so... it's not been called! (again, please test! If it does fail then I accept that I might be wrong here)
- Digging deeper, the actually invoked ctrl is that of RDG and not -- as we expected -- TSG.
- Probably, TLS is not running on top of TSG, but on top of RDG. But I do not know about these details.

Reply:
Apparently, TSG/RPC transport operates in blocking mode during tls_do_handshake() call (in tls.c), so whole handshake is done after single BIO_do_handshake() and the statement with BIO_get_fd() is not even reached.
I guess, that means BIO_C_GET_FD implementation in RPC can be omitted for now...